### PR TITLE
[typed] Introduce typed mode ("use types")

### PR DESCRIPTION
**Pull request by rossberg**
@arv, PTAL

I created the typed-mode branch in my fork for now.

### CODE REVIEW

File: src/staticsemantics/LanguageMode.js [hunk 1]
```diff
@@ -16,19 +16,24 @@
...
 
 export function findLanguageMode(statements, mode) {
   if (mode !== STRONG_MODE) {
     for (let i = 0; i < statements.length; i++) {
       if (!statements[i].isDirectivePrologue()) {
         break;
-      }
-      if (statements[i].isUseStrictDirective()) {
-        if (mode === SLOPPY_MODE) {
+      } else if (statements[i].isUseStrictDirective()) {
```
Comment (arv) [inline src/staticsemantics/LanguageMode.js]: no need for else after break/continue/return/throw

**Comment by arv** [comment]
Very nice. Very excited to see this starting.

File: src/staticsemantics/LanguageMode.js (hunk 1 above)
Comment (rossberg) [inline src/staticsemantics/LanguageMode.js]: Done. (I liked how it's more symmetric and similar to checkDirective. :) )

**Comment by rossberg** [comment]
PTAL

**Comment by arv** [comment]
LGTM

**Comment by arv** [comment]
Do you want me to merge it into google/traceur-compiler/strong-mode or did you want to add more logic to this before landing?

**Comment by rossberg** [comment]
Upstreaming this makes sense, but perhaps we should keep it in a branch separate from strong-mode for now? (like the typed-mode I introduced in my fork.)

**Comment by arv** [comment]
I'm fine with that. Maintaining branches have some overhead but we can
merge the strong and typed branches when we have some more logic
implemented.

On Mon, Jun 8, 2015 at 12:59 PM, rossberg-chromium <notifications@github.com

> wrote:
> 
> Upstreaming this makes sense, but perhaps we should keep it in a branch
> separate from strong-mode for now? (like the typed-mode I introduced in my
> fork.)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/rossberg-chromium/traceur-soundscript/pull/1#issuecomment-110073288
> .

## 

erik

**Pull request merged**